File: compiler_Haskell/src/ConvertGlobal.hs
module ConvertGlobal where

import Data.Map as M
import Data.Set as S
import RunRun
import KNormal as K
import Control.Monad.State
import qualified Alpha (g)


convertGlobal :: K -> RunRun K
convertGlobal e = do
--        eprint =<< (globals <$> get)
        eprint e
        eputstrln "convering ..."
        find_and_convert e


find_and_convert :: K -> RunRun K
find_and_convert (KLetRec f@(KFunc {kname=(x,t),kargs=yts,kbody=_}) e2) = do
           -- eputstrln $ "converting inside " ++ x
           -- eputstrln "  "
            e1' <- convert_each f =<< globals <$> get
           -- eputstrln "  "
           -- eputstrln "globals are"
           -- eputstrln "  "
           -- gb <- globals <$> get
           -- eprint gb
           -- eputstrln "  "
           -- eprint e1
           -- eputstrln " "
           -- eputstrln " to "
           -- eputstrln " "
           -- eprint e1'
            e2' <- find_and_convert e2
            return $ KLetRec (KFunc {kname=(x,t),kargs=yts,kbody=e1'}) e2'
find_and_convert (Let xt e1 e2) =
            Let xt <$> find_and_convert e1 <*> find_and_convert e2
find_and_convert (LetTuple xts y e) =
            LetTuple xts y <$> find_and_convert e
find_and_convert (If x e1 e2) =
            If x <$> find_and_convert e1 <*> find_and_convert e2
find_and_convert (FIfCmp cmp x y e1 e2) =
            FIfCmp cmp x y <$> find_and_convert e1 <*> find_and_convert e2
find_and_convert e = return e


convert_each :: KFundef -> Map String Global -> RunRun K
convert_each (K.KFunc { K.kname = (_,_), K.kargs = _, K.kbody = e1}) global = do
        mapping <- sequence (mapWithKey (\k _ -> genid k) global)
--        eprint mapping
        let newglobal = M.mapKeys (mapping M.!) global
        e1' <- Alpha.g mapping e1
        return $ convert newglobal e1'


convert :: Map String Global -> K -> K
convert genv (Let xt e1 e2) =
                    let (fs,ds) = free genv e1 in
                    insertlet fs (Let xt e1 (convert ds e2))
convert genv (LetTuple xts y e) =
                    case M.lookup y genv of
                            Just global_y -> let e' = convert (M.delete y genv) e in
                                             insertlet (M.singleton y global_y) (LetTuple xts y e')
                            Nothing       -> let e' = convert genv e in
                                             LetTuple xts y e'
convert genv (KLetRec f e2) = KLetRec f (convert genv e2)
convert genv e = let (fenv,_) = free genv e in
                 insertlet fenv e


free :: Map String Global -> K -> (Map String Global, Map String Global)
free genv e = let fve = fv e in
              let keysetGenv = keysSet genv in
              let denv = S.foldl (\a b -> M.delete b a) genv (S.intersection fve keysetGenv) in
              let fenv = S.foldl (\a b -> M.delete b a) genv (keysetGenv S.\\ fve) in
              (fenv, denv)


insertlet :: Map String Global -> K -> K
insertlet fenv e =
        foldlWithKey (\e' x (Global{typ=t,addr=p}) ->
                Let (x,t) (Int p) e') e fenv

File: min-caml/convert_global.ml
open KNormal


let free map e =
  let fve = fv e in
  let keysetMap = M.fold (fun k a set -> S.add k set) map S.empty in
  let denv = S.fold (fun k m -> M.remove k m) (S.inter fve keysetMap) map in
  let fenv = S.fold (fun k m -> M.remove k m) (S.diff  keysetMap fve) map in
  (fenv, denv)


let insertlet fenv e = M.fold (fun x {Global.typ=t; Global.addr=p} e' -> Let ((x,t), (Int p), e')) fenv e


let rec convert map e =
  match e with
  | Let (xt, e1, e2) ->
      let (fs,ds) = free map e1 in
      insertlet fs (Let (xt, e1, (convert ds e2)))
  | LetTuple (xts, y, e) ->
      if M.mem y map then
        let e' = convert (M.remove y map) e in
        insertlet (M.singleton y (M.find y map)) (LetTuple (xts, y, e'))
      else
        let e' = convert map e in
        LetTuple (xts, y, e')
  | LetRec (f, e2) -> LetRec (f, convert map e2)
  | e -> let (fenv,_) = free map e in insertlet fenv e




let convert_each e1 global =
  let mapping = M.mapi (fun k _ -> Id.genid k) global in
  let newglobal = M.fold (fun k k' map -> M.add k' (M.find k global) map) mapping M.empty in
  let e1 = Alpha.g mapping e1 in
  convert newglobal e1



let rec find_and_convert exp =
  match exp with
  | LetRec ({name = xt; args=yts; body=e1}, e2) ->
      let e1 = convert_each e1 !Global.global in
      let e2 = find_and_convert e2 in
      LetRec ({name=xt; args=yts; body=e1}, e2)
  | Let (xt, e1, e2) ->
      Let (xt, find_and_convert e1, find_and_convert e2)
  | LetTuple (xts, y, e) ->
      LetTuple (xts, y, find_and_convert e)
  | IfFZero (x, e1, e2) ->
      IfFZero (x, find_and_convert e1, find_and_convert e2)
  | IfEq (x,y,e1,e2) ->
      IfEq (x,y, find_and_convert e1, find_and_convert e2)
  | IfLE (x,y,e1,e2) ->
      IfLE (x,y, find_and_convert e1, find_and_convert e2)
  | IfFLt (x,y,e1,e2) ->
      IfFLt (x,y, find_and_convert e1, find_and_convert e2)
  | e -> e

let f = find_and_convert
